edgelet/support-bundle/src/support_bundle.rs

```diff
@@ -0,0 +1,785 @@
+// Copyright (c) Microsoft. All rights reserved.
+
+use std::env;
+use std::ffi::OsString;
+use std::fs::File;
+use std::io::{Cursor, Seek};
+use std::path::Path;
+use std::process::Command as ShellCommand;
+
+use chrono::{DateTime, Local, NaiveDateTime, Utc};
+use failure::Fail;
+use futures::{Future, Stream};
+use tokio::prelude::*;
+use zip::{write::FileOptions, CompressionMethod, ZipWriter};
+
+use edgelet_core::{LogOptions, LogTail, Module, ModuleRuntime};
+
+use crate::error::{Error, ErrorKind};
+use crate::logs::pull_logs;
+
+pub fn make_bundle<M>(
+    output_location: OutputLocation,
+    log_options: LogOptions,
+    include_ms_only: bool,
+    verbose: bool,
+    iothub_hostname: Option<String>,
+    runtime: M,
+) -> Box<dyn Future<Item = (Box<dyn Read + Send>, u64), Error = Error> + Send>
```

Comment:
Do you need to make a trait object here? `-> impl Future<Item = (impl Read + Send, u64), Error = Error>` should work too, and then you wouldn't need to `Box` the `Future` nor the `Read` either.

Reply:
It has to be boxed since there are 2 match arms that return different futures

Reply:
Ah yes, and two different `Read` impls as well.